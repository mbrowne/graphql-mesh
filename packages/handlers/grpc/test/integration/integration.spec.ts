import { join } from 'path';
import { GraphQLSchema, printSchema, validateSchema } from 'graphql';
import lodashGet from 'lodash.get';
import InMemoryLRUCache from '@graphql-mesh/cache-localforage';
import { InMemoryStoreStorageAdapter, MeshStore } from '@graphql-mesh/store';
import type { KeyValueCache, YamlConfig } from '@graphql-mesh/types';
import { defaultImportFn, PubSub } from '@graphql-mesh/utils';
import { DefaultLogger } from '@graphql-mesh/utils';
import {
  loadPackageDefinition,
  Server,
  ServerCredentials,
  ServiceClientConstructor,
} from '@grpc/grpc-js';
import { load } from '@grpc/proto-loader';
import { fetch as fetchFn } from '@whatwg-node/fetch';
import GrpcHandler from '../../src/index.js';

const wrapServerWithReflection = require('grpc-node-server-reflection').default;

const TEST_GRPC_SERVER_PORT = process.env.TEST_GRPC_SERVER_PORT || 50051;

describe('gRPC Handler integration tests', () => {
  let cache: KeyValueCache;
  let pubsub: PubSub;
  let store: MeshStore;
  let logger: DefaultLogger;
  beforeEach(() => {
    cache = new InMemoryLRUCache();
    pubsub = new PubSub();
    store = new MeshStore('grpc-test', new InMemoryStoreStorageAdapter(), {
      readonly: false,
      validate: false,
    });
    logger = new DefaultLogger('grpc-test');
  });
  afterEach(() => {
    pubsub.publish('destroy', undefined);
  });

  test('Load protobuf via reflection', async () => {
    // load protobuf from two different packages
    const server = await startGrpcServer({
      'comments.proto': { servicePath: 'foo.SampleService' },
      'empty.proto': { servicePath: 'io.xtech.Example' },
    });

    const config: YamlConfig.GrpcHandler = {
      endpoint: 'localhost:' + TEST_GRPC_SERVER_PORT,
    };

    const handler = new GrpcHandler({
      name: Date.now().toString(),
      config,
      cache,
      pubsub,
      store,
      logger,
      importFn: defaultImportFn,
      baseDir: __dirname,
    });

    const { schema } = await handler.getMeshSource({ fetchFn });

    expect(schema).toBeInstanceOf(GraphQLSchema);
    expect(validateSchema(schema)).toHaveLength(0);
    expect(printSchema(schema)).toMatchSnapshot();

    return new Promise<void>(resolve => {
      server.tryShutdown(err => {
        if (err) {
          server.forceShutdown();
        }
        resolve();
      });
    });
  });
});

type ProtoMap = {
  [protoFilePath: string]: {
    servicePath: string;
  };
};

async function startGrpcServer(protoMap: ProtoMap) {
  const server: Server = wrapServerWithReflection(new Server());

  for (const [protoFilePath, { servicePath }] of Object.entries(protoMap)) {
    const packageDefinition = await load(protoFilePath, {
      includeDirs: [join(__dirname, '../fixtures/proto-tests')],
    });
    const grpcObject = loadPackageDefinition(packageDefinition);
    server.addService((lodashGet(grpcObject, servicePath) as ServiceClientConstructor).service, {});
  }

  return new Promise<Server>((resolve, reject) => {
    server.bindAsync(
      'localhost:' + TEST_GRPC_SERVER_PORT,
      ServerCredentials.createInsecure(),
      (error, port) => {
        if (error) {
          reject(error);
          return;
        }
        server.start();

        console.log('gRPC Server started, listening: localhost:' + port);
        resolve(server);
      },
    );
  });
}
