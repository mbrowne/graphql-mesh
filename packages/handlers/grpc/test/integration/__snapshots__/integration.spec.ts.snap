// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gRPC Handler integration tests Load protobuf via reflection 1`] = `
"directive @grpcMethod(rootJsonName: String, objPath: String, methodName: String, responseStream: Boolean) on FIELD_DEFINITION

directive @grpcConnectivityState(rootJsonName: String, objPath: String) on FIELD_DEFINITION

directive @grpcRootJson(name: String, rootJson: ObjMap, loadOptions: ObjMap) repeatable on OBJECT

type Query {
  foo_SampleService_GetSample(input: foo_GetSampleRequest_Input): foo_SampleResponse
  foo_SampleService_connectivityState(tryToConnect: Boolean): ConnectivityState
  io_xtech_Example_GetMovies(input: io_xtech_MovieRequest_Input): io_xtech_MoviesResult
  io_xtech_Example_GetEmpty(input: io_xtech_MovieRequest_Input): io_xtech_EmptyRequest
  io_xtech_Example_SearchMoviesByCast(input: io_xtech_SearchByCastRequest_Input): [io_xtech_Movie]
  io_xtech_Example_connectivityState(tryToConnect: Boolean): ConnectivityState
  io_xtech_AnotherExample_GetMovies(input: io_xtech_MovieRequest_Input): io_xtech_MoviesResult
  io_xtech_AnotherExample_SearchMoviesByCast(input: io_xtech_SearchByCastRequest_Input): [io_xtech_Movie]
  io_xtech_AnotherExample_connectivityState(tryToConnect: Boolean): ConnectivityState
}

type foo_SampleResponse {
  sampleId: String
}

input foo_GetSampleRequest_Input {
  sampleId: String
}

enum ConnectivityState {
  IDLE
  CONNECTING
  READY
  TRANSIENT_FAILURE
  SHUTDOWN
}

type io_xtech_MoviesResult {
  result: [io_xtech_Movie]
}

type io_xtech_Movie {
  name: String
  year: Int
  rating: Float
  cast: [String]
  time: google_protobuf_Timestamp
  genre: io_xtech_Genre
}

type google_protobuf_Timestamp {
  seconds: BigInt
  nanos: Int
}

"""
The \`BigInt\` scalar type represents non-fractional signed whole numeric values.
"""
scalar BigInt

enum io_xtech_Genre {
  UNSPECIFIED
  ACTION
  DRAMA
}

input io_xtech_MovieRequest_Input {
  movie: io_xtech_Movie_Input
}

input io_xtech_Movie_Input {
  name: String
  year: Int
  rating: Float
  cast: [String]
  time: google_protobuf_Timestamp_Input
  genre: io_xtech_Genre
}

input google_protobuf_Timestamp_Input {
  seconds: BigInt
  nanos: Int
}

scalar io_xtech_EmptyRequest @specifiedBy(url: "http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf")

input io_xtech_SearchByCastRequest_Input {
  castName: String
}

type Mutation {
  foo_SampleService_CreateSample(input: foo_CreateSampleRequest_Input): foo_SampleResponse
  foo_SampleService_UpdateSample(input: foo_UpdateSampleRequest_Input): foo_SampleResponse
  foo_SampleService_DeleteSample(input: foo_DeleteSampleRequest_Input): foo_SampleResponse
}

input foo_CreateSampleRequest_Input {
  description: String
  type: String
}

input foo_UpdateSampleRequest_Input {
  sampleId: String
}

input foo_DeleteSampleRequest_Input {
  sampleId: String
}

scalar ObjMap"
`;
